### PR TITLE
Feature - human in the loop

### PR DESCRIPTION
**Pull request by damntrecky**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

## Description

- Add a feature to allow user to upload dependencies to resolve build errors during the transformation

## Checklist
- [ ] My code follows the code style of this project
- [x ] I have added tests to cover my changes
- [ ] A short description of the change has been added to the **[CHANGELOG](https://github.com/aws/aws-toolkit-vscode/blob/master/CONTRIBUTING.md)** if the change is customer-facing in the IDE.
- [ x] I have added metrics for my changes (if required)
 
## License
I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/testE2E/amazonqGumby/transformByQ.test.ts [hunk 1]
```diff
@@ -13,7 +13,7 @@ import { getSha256, uploadArtifactToS3, zipCode } from '../../codewhisperer/serv
 import request from '../../common/request'
```
Comment (damntrecky) [inline packages/core/src/testE2E/amazonqGumby/transformByQ.test.ts:13]: TODO: Move this into amazonQGumby test folder

File: packages/core/src/test/amazonqGumby/humanInTheLoopHandler.test.ts [hunk 2]
```diff
@@ -0,0 +1,60 @@
+/*!
```
Comment (damntrecky) [inline packages/core/src/test/amazonqGumby/humanInTheLoopHandler.test.ts:1]: TODO: Refactor/rename this test file since we have a separate HumanInTheLoopManager

File: packages/core/scripts/build/copyFiles.ts [hunk 3]
```diff
@@ -27,7 +27,7 @@ interface CopyTask {
 const tasks: CopyTask[] = [
     { target: path.join('src', 'templates') },
     { target: path.join('src', 'test', 'shared', 'cloudformation', 'yaml') },
-    { target: path.join('src', 'test', 'codewhisperer', 'service', 'resources') },
+    { target: path.join('src', 'test', 'amazonqGumby') },
```
Comment (damntrecky) [inline packages/core/scripts/build/copyFiles.ts:30]: TODO: Verify all Gumby tests are outside of the codewhisperer

File: packages/core/src/test/codewhisperer/service/transformByQ/humanInTheLoopHandler.test.ts [hunk 4]
```diff
@@ -0,0 +1,150 @@
+/*!
```
Comment (damntrecky) [inline packages/core/src/test/codewhisperer/service/transformByQ/humanInTheLoopHandler.test.ts:1]: TODO: Rename this file since we have a humanInTheLoop manager class. These should be refactored into a utilities file test or something.

File: packages/core/src/codewhisperer/service/transformByQ/transformationHubViewProvider.ts [hunk 5]
```diff
@@ -308,7 +308,7 @@ export class TransformationHubViewProvider implements vscode.WebviewViewProvider
 
         let planSteps = transformByQState.getPlanSteps()
         if (jobPlanProgress['generatePlan'] === StepProgress.Succeeded && transformByQState.isRunning()) {
-            planSteps = await getTransformationSteps(transformByQState.getJobId())
+            planSteps = await getTransformationSteps(transformByQState.getJobId(), false)
```
Comment (damntrecky) [inline packages/core/src/codewhisperer/service/transformByQ/transformationHubViewProvider.ts:311]: Question: Not sure the false flag is necessary.

File: packages/core/src/codewhisperer/models/model.ts [hunk 6]
```diff
@@ -318,9 +319,31 @@ export class ZipManifest {
     dependenciesRoot: string | undefined = 'dependencies/'
     buildLogs: string = 'build-logs.txt'
     version: string = '1.0'
+    hilCapabilities: string[] = ['HIL_1pDependency_VersionUpgrade']
```
Comment (damntrecky) [inline packages/core/src/codewhisperer/models/model.ts:322]: Note: Passing along this value in the IDE is how we "feature flag". Seeing this capability will trigger API to send new functionality to API.

Backend changes are already in production.

File: packages/core/src/codewhisperer/client/codewhisperer.ts [hunk 7]
```diff
@@ -38,8 +38,8 @@ export interface CodeWhispererConfig {
 }
 
 export const defaultServiceConfig: CodeWhispererConfig = {
-    region: 'us-east-1',
-    endpoint: 'https://codewhisperer.us-east-1.amazonaws.com/',
+    region: 'us-west-2',
```
Comment (damntrecky) [inline packages/core/src/codewhisperer/client/codewhisperer.ts:41]: BLOCKER: Revert this change

File: packages/core/src/test/amazonqGumby/humanInTheLoopManager.test.ts [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+        const uploadFolderInfo = HumanInTheLoopManager.instance.getUploadFolderInfo()
+        assert.strictEqual(uploadFolderInfo.name, 'q-pom-dependency-update')
+        assert.strictEqual(uploadFolderInfo.path, HumanInTheLoopManager.instance.getUserDependencyUpdateDir())
+    })
+    it('will getCompileDependencyListFolderInfo()', async function () {
+        const compileFolderInfo = HumanInTheLoopManager.instance.getCompileDependencyListFolderInfo()
+        assert.strictEqual(compileFolderInfo.name, 'q-pom-dependency-list')
+        assert.strictEqual(compileFolderInfo.path, HumanInTheLoopManager.instance.getTmpDependencyListDir())
+    })
+    it.skip('will createPomFileCopy() and delete artifact', async function () {
```
Comment (damntrecky) [inline packages/core/src/test/amazonqGumby/humanInTheLoopManager.test.ts:28]: Note: Added a test skip until I can resolve this. I tried adding the `normalize` and the `assertEqualPaths` but the result is one filename has a prefix with a `\`. 

I will resolve this on a separate branch and use a drafted PR to check CI until this passes.

File: packages/amazonq/.changes/next-release/Feature-783a661b-705e-403f-af64-76db8c209496.json [hunk 9]
```diff
@@ -0,0 +1,4 @@
+{
+	"type": "Feature",
+	"description": "Human in the loop - Adding human intervention to help update dependencies during the Amazon Q Transformation process\""
```
[the remark this answers is not in the document]
Comment (damntrecky) [reply packages/amazonq/.changes/next-release/Feature-783a661b-705e-403f-af64-76db8c209496.json:3]: Not sure. I think this gets manually processed/curated during the deployment and could be adjusted there?

File: packages/core/scripts/build/copyFiles.ts [hunk 10]
```diff
@@ -27,7 +27,7 @@ interface CopyTask {
 const tasks: CopyTask[] = [
     { target: path.join('src', 'templates') },
     { target: path.join('src', 'test', 'shared', 'cloudformation', 'yaml') },
-    { target: path.join('src', 'test', 'codewhisperer', 'service', 'resources') },
```
Comment (hayemaxi) [inline packages/core/scripts/build/copyFiles.ts:30]: Do we need to delete this folder?

File: packages/core/scripts/build/copyFiles.ts (hunk 3 above)
Comment (hayemaxi) [inline packages/core/scripts/build/copyFiles.ts:30]: why are we copying the whole test folder with .ts files? Can we only copy the resources inside?

File: packages/core/src/amazonqGumby/app.ts [hunk 11]
```diff
@@ -27,6 +27,10 @@ export function init(appContext: AmazonQAppInitContext) {
         transformationFinished: new vscode.EventEmitter<any>(),
         processHumanChatMessage: new vscode.EventEmitter<any>(),
         linkClicked: new vscode.EventEmitter<any>(),
+        startHumanInTheLoopIntervention: new vscode.EventEmitter<any>(),
+        promptForDependencyHIL: new vscode.EventEmitter<any>(),
+        HILSelectionUploaded: new vscode.EventEmitter<any>(),
```
Comment (justinmk3) [inline packages/core/src/amazonqGumby/app.ts:32]: why is HIL abbreviated only sometimes? also, `HILSelectionUploaded` does not follow the naming conventions.
```suggestion
        startHilIntervention: new vscode.EventEmitter<any>(),
        promptForDependencyHil: new vscode.EventEmitter<any>(),
        hilSelectionUploaded: new vscode.EventEmitter<any>(),
```

File: packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+import * as vscode from 'vscode'
+import * as os from 'os'
+import path from 'path'
+import { FolderInfo } from '../../models/model'
+import { fsCommon } from '../../../srcShared/fs'
+import { createPomCopy, replacePomVersion } from './transformFileHandler'
+import { IManifestFile } from '../../../amazonqFeatureDev/models'
+import { getLogger } from '../../../shared/logger'
+
+export class HumanInTheLoopManager {
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts:14]: what does this class do? needs a docstring

File: packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+        try {
+            await fsCommon.delete(this.tmpDependencyListDir)
+        } catch (e: any) {
+            getLogger().error(artifactCleanUpErrorMessage(e))
+        }
+        try {
+            await fsCommon.delete(this.tmpDownloadsDir)
+        } catch (e: any) {
+            getLogger().error(artifactCleanUpErrorMessage(e))
+        }
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts:86]: this could be a single try/catch

File: packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts [hunk 14]
```diff
@@ -124,6 +138,41 @@ export async function uploadArtifactToS3(
...
+                codeTransformRequestId: response.$response.requestId,
+                result: MetadataResult.Pass,
+            })
+            // always store request ID, but it will only show up in a notification if an error occurs
+            return response.transformationStatus
+        }
+    } catch (e: any) {
+        const errorMessage = (e as Error).message
+        getLogger().error(`CodeTransformation: ResumeTransformation error = ${errorMessage}`)
+        telemetry.codeTransform_logApiError.emit({
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts:163]: why is this a separate metric? the `codeTransform_logApiLatency` metric could have a result=fail with `reason` field set. there is no need to have separate "failure" metrics for each event.

File: packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts [hunk 15]
```diff
@@ -328,11 +343,38 @@ export class Messenger {
         )
     }
 
+    public sendKnownErrorResponse(type: ErrorResponseType, tabID: string) {
```
Comment (hayemaxi) [inline packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts:346]: the function above has a nice docstring, can we add one here that explains this one as well?

File: packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts [hunk 16]
```diff
@@ -206,3 +204,38 @@ export async function getVersionData() {
...
+    let environment = process.env
+    // if JAVA_HOME not found or not matching project JDK, get user input for it and set here
+    if (transformByQState.getJavaHome() !== undefined) {
+        environment = { ...process.env, JAVA_HOME: transformByQState.getJavaHome() }
+    }
+
+    const spawnResult = spawnSync(baseCommand, args, {
+        // default behavior is looks for pom.xml in this root
+        cwd: dependenciesFolder.path,
+        shell: true,
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts:230]: are you sure that shell is needed? it can cause side effects depending on the user's shell config. if it's needed, a comment should explain why

File: packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts [hunk 17]
```diff
@@ -220,8 +224,10 @@ export class Messenger {
         )
     }
 
-    public sendJobSubmittedMessage(tabID: string, disableJobActions: boolean = false) {
-        const message = CodeWhispererConstants.jobStartedChatMessage
+    public sendJobSubmittedMessage(tabID: string, disableJobActions: boolean = false, message?: string) {
+        if (!message) {
+            message = CodeWhispererConstants.jobStartedChatMessage
```
Comment (hayemaxi) [inline packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts:229]: nit: set default in function signature

File: packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json [hunk 18]
```diff
@@ -0,0 +1,7 @@
+{
```
Comment (justinmk3) [inline packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json:1]: test "resources" should live in testFixtures https://github.com/aws/aws-toolkit-vscode/tree/f9d94712dfbb6a6be5597be72f2f5201a7913c52/packages/core/src/testFixtures

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 19]
```diff
@@ -194,6 +266,214 @@ export async function preTransformationUploadCode() {
...
+        } finally {
+            transformByQState.getChatControllers()?.errorThrown.fire({
+                error: err,
+                tabID: ChatSessionManager.Instance.getSession().tabID,
+            })
+        }
+        CodeTransformTelemetryState.instance.setCodeTransformMetaDataField({
+            errorMessage: err.message,
+        })
+        telemetry.codeTransform_humanInTheLoop.emit({
```
Comment (hayemaxi) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:360]: We wont submit telemetry if the nested try/catch above throws. Is this intended/possible?
Consider wrapping this function in telemtry.run instead of emitting once at the end.

Comment (hayemaxi) [reply packages/core/src/codewhisperer/commands/startTransformByQ.ts:360]: I think it would make more sense to move all of the code in this catch block (except `await terminate` in the inner block) to the outer finally block.

File: packages/core/src/test/amazonqGumby/humanInTheLoopManager.test.ts (hunk 8 above)
Comment (hayemaxi) [reply packages/core/src/test/amazonqGumby/humanInTheLoopManager.test.ts:28]: would be good to add this as a comment, so folks who read the code can see why its skipped

File: packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts (hunk 13 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts:86]: `fsCommon.delete` throws an error no? We want the minimal impact for artifact deletion if something went wrong.

File: packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts (hunk 14 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/service/transformByQ/transformApiHandler.ts:163]: This is a standing pattern. Closing out.

File: packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts (hunk 12 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/service/transformByQ/humanInTheLoopManager.ts:14]: Helps encapsulate all the interactions for artifact management during this feature. Makes it easier to cleanup

File: packages/core/scripts/build/copyFiles.ts (hunk 3 above)
Comment (damntrecky) [reply packages/core/scripts/build/copyFiles.ts:30]: Yes we can, I missed the resources folder copy.

File: packages/core/scripts/build/copyFiles.ts (hunk 10 above)
Comment (damntrecky) [reply packages/core/scripts/build/copyFiles.ts:30]: Its not used anymore. So yes.

File: packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts (hunk 15 above)
Comment (damntrecky) [reply packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts:346]: @JuliaABurch Can we add this?

File: packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts (hunk 16 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts:230]: Historically you and @dhasani23 have talked about this. The most context I have is its needed to help find executable path.

But I have also since questioned if its needed now vs a couple months ago.

File: packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json (hunk 18 above)
Comment (damntrecky) [reply packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json:1]: Is there a functional reason for this?

We are trying to keep our fixtures and resources close to our code since upstream refactors have happened multiple times the last couple months. For risk mitigation, we want to keep our resources/code away from shared spaces when possible.

File: packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts (hunk 16 above)
Comment (damntrecky) [reply packages/core/src/codewhisperer/service/transformByQ/transformMavenHandler.ts:230]: We have an appsec finding on this and will follow up on this to close by 6/26.

**Review by mk-fan — APPROVED** [review]

File: packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts (hunk 15 above)
Comment (JuliaABurch) [reply packages/core/src/amazonqGumby/chat/controller/messenger/messenger.ts:346]: /**
     * This method renders an error message as a plain message with no other prompt or action
     * for the user to follow. Either the job can continue and this message is purely for
     * informational purposes, or some other error workflow is meant to contribute a 
     * follow-up with a user action.
     */

**Review by JuliaABurch — APPROVED** [review]

File: packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json (hunk 18 above)
Comment (hayemaxi) [reply packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json:1]: What's the risk here? If a refactor breaks your tests... we will know because the tests fail.  Not sure if there is a technical reason though.

File: packages/core/scripts/build/copyFiles.ts (hunk 10 above)
Comment (hayemaxi) [reply packages/core/scripts/build/copyFiles.ts:30]: then can we delete it in this PR before its forgotten about? Looks like you eliminated the last use case.

File: packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json (hunk 18 above)
Comment (damntrecky) [reply packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json:1]: The risk is having changes to our codebase without no visibility. This has happened several times in the past. 

Yet, the inverse is not true. We cannot make changes to the codebase without your team approval.

Comment (damntrecky) [reply packages/core/src/test/amazonqGumby/resources/files/humanInTheLoop/downloadResults/manifest.json:1]: The risk is having changes to our codebase without no visibility. This has happened several times in the past. 

Yet, the inverse is not true. We cannot make changes to the codebase without your team approval.

**Review by awsanakkala — APPROVED** [review]

File: packages/core/src/codewhisperer/commands/startTransformByQ.ts [hunk 20]
```diff
@@ -194,6 +266,208 @@ export async function preTransformationUploadCode() {
...
+        return true
+    } finally {
+        await sleep(1000)
+        telemetry.codeTransform_humanInTheLoop.emit({
+            codeTransformSessionId: CodeTransformTelemetryState.instance.getSessionId(),
+            codeTransformJobId: jobId,
+            codeTransformMetadata: CodeTransformTelemetryState.instance.getCodeTransformMetaDataString(),
+            result: MetadataResult.Fail,
+            // TODO: make a generic reason field for telemetry logging so we don't log sensitive PII data
+            reason: 'Runtime error occurred',
```
Comment (justinmk3) [inline packages/core/src/codewhisperer/commands/startTransformByQ.ts:370]: @damntrecky the default command error handling (and `telemetry.run()`) now sets the error message in the `reasonDesc` field, since https://github.com/aws/aws-toolkit-vscode/pull/5158 . And https://github.com/aws/aws-toolkit-vscode/pull/5215 removes potential PII from the error message.

- `reason` and `reasonDesc` are explained [here](https://github.com/aws/aws-toolkit-common/blob/867fff1ad266088443b1ad2f2008153a2560e758/telemetry/definitions/commonDefinitions.json#L1263-L1272).
- [getTelemetryReasonDesc](https://github.com/aws/aws-toolkit-vscode/blob/059facab1c3bf77c16291d30847d606b16330739/packages/core/src/shared/errors.ts#L316) can/should be used to set the `reasonDesc` field, for code not implemented as a vscode command and not using `telemetry.run()`.
